crates/subspace-archiving/src/archiver.rs

```diff
@@ -614,56 +631,82 @@ impl Archiver {
-                }
-                | SegmentItem::BlockContinuation {
-                    bytes,
-                    object_mapping,
-                } => {
-                    for block_object in object_mapping.objects() {
-                        // `+1` corresponds to `SegmentItem::X {}` enum variant encoding
-                        let offset_in_segment = base_offset_in_segment
-                            + 1
-                            + Compact::compact_len(&(bytes.len() as u32))
-                            + block_object.offset as usize;
-                        let raw_piece_offset = (offset_in_segment % RawRecord::SIZE)
-                            .try_into()
-                            .expect("Offset within piece should always fit in 32-bit integer; qed");
-                        corrected_object_mapping.push(GlobalObject {
-                            hash: block_object.hash,
-                            piece_index: source_piece_indexes[offset_in_segment / RawRecord::SIZE],
-                            offset: raw_piece_offset,
-                        });
-                    }
-                }
-                SegmentItem::ParentSegmentHeader(_) => {
-                    // Ignore, no objects mappings here
+        match last_item {
+            SegmentItem::Padding => {
+                unreachable!("Segment during archiving never contains SegmentItem::Padding; qed");
+            }
+            SegmentItem::Block {
+                bytes,
+                object_mapping,
+            }
+            | SegmentItem::BlockStart {
+                bytes,
+                object_mapping,
+            }
+            | SegmentItem::BlockContinuation {
+                bytes,
+                object_mapping,
+            } => {
+                for block_object in object_mapping.objects() {
+                    // `+1` corresponds to `SegmentItem::X {}` enum variant encoding
+                    let offset_in_segment = base_offset_in_segment
+                        + 1
+                        + Compact::compact_len(&(bytes.len() as u32))
+                        + block_object.offset as usize;
+                    let raw_piece_offset = (offset_in_segment % RawRecord::SIZE)
+                        .try_into()
+                        .expect("Offset within piece should always fit in 32-bit integer; qed");
+                    corrected_object_mapping.push(GlobalObject {
+                        hash: block_object.hash,
+                        piece_index: source_piece_indexes[offset_in_segment / RawRecord::SIZE],
+                        offset: raw_piece_offset,
+                    });
                 }
             }
-
-            base_offset_in_segment += segment_item.encoded_size();
+            SegmentItem::ParentSegmentHeader(_) => {
+                // If this block filled the last segment, then the last item in the next segment
+                // will be a header, with no mappings.
```

Comment:
I like the old comment better because it clearly states that segment headers can't/don't contain object mappings.

The description of "this block" (which does not exist in the context of this function) requires a lot of brain cycles to process and ends with `will be a header, with no mappings`, basically inviting an idea that under other circumstances there might be a header with mappings, which is not true, headers never have mappings regardless of surrounding context.

Reply:
I think we also need to explain why this isn’t unreachable, how about:
```suggestion
                // Headers never contain object mappings. They can only be the last item when
                // the latest block filled the last segment exactly.
```

Reply:
It was reachable with previous implementation and after reading above comment with the context of potentially unnecessarily changed `with_initial_state()` API I'm not 100% sure we need to constrain this function to only process one segment item.

Reply:
In the latest code in this branch, I don't think there's any need to call this function on more than one segment item.

This function is now called after every block is added, and for the `BlockContinuation` added by `with_initial_state()`. It runs until segments are exhausted, and then for any trailing `BlockContinuation`.

Can you find a case where a block adds more than one item to a segment?